soc/riscv/riscv-privileged/common/soc_common_irq.c

```diff
@@ -36,21 +36,25 @@ void z_riscv_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flag
 	riscv_clic_irq_priority_set(irq, prio, flags);
 }
 
-#else /* PLIC + HLINT/CLINT or HLINT/CLINT only */
+#else /* PLIC + HLINT/CLINT or HLINT/CLINT or custom interrupt controller */
 
 void arch_irq_enable(unsigned int irq)
 {
 	uint32_t mie;
 
-#if defined(CONFIG_RISCV_HAS_PLIC)
+#if defined(CONFIG_RISCV_HAS_PLIC) || defined(CONFIG_RISCV_HAS_CUSTOM_IC)
 	unsigned int level = irq_get_level(irq);
 
 	if (level == 2) {
 		irq = irq_from_level_2(irq);
+#if defined(CONFIG_RISCV_HAS_PLIC)
 		riscv_plic_irq_enable(irq);
+#else
+		riscv_custom_ic_irq_enable(irq);
```

Comment:
This is wrong because you are already assuming a lot of things about the custom interrupt controller, for example that it has multiple levels. If the platform has a custom interrupt controller, `arch_irq_enable` must only call into `z_soc_`, nothing else. If the platform needs more bits and pieces (like multi-level IRQ management and such) that must be dealt with in the SOC specific code.

Reply:
Oh, I think I've got your idea.
So for the soc_common_irq level you just suggest to redirect the calls to soc-specific implementation and not to reuse PLIC path in any way?

Reply:
correct

Reply:
Hi @carlocaione ,

Done! Please take a look.

I still have doubts about:
1) RISCV_CUSTOM_INTERRUPT_CONTROLLER (ARM way) vs RISCV_HAS_CUSTOM_IC (RISCV way)
2) should we have z_soc_custom_ic_ prefix instead of z_soc_ prefix
3) dependence from RISC-V privileged: if we remove this dependency, should we create and move the code to e.g. soc/riscv/common directory ?

Could you please share your thoughts regarding the points above?
Thanks!

Reply:
> 1. RISCV_CUSTOM_INTERRUPT_CONTROLLER (ARM way) vs RISCV_HAS_CUSTOM_IC (RISCV way)

`CONFIG_RISCV_CUSTOM_INTERRUPT_CONTROLLER`

> 2. should we have z_soc_custom_ic_ prefix instead of z_soc_ prefix

`z_soc_*`

> 3. dependence from RISC-V privileged: if we remove this dependency, should we create and move the code to e.g. soc/riscv/common directory ?

This is not related to the RISC-V privileged, so there should be no dependency.

Reply:
>This is not related to the RISC-V privileged, so there should be no dependency.
Agree. But what about the code? Should we move from soc/riscv/riscv-privileged/common/soc_common_irq.c to a new directory soc/riscv/common/soc_common_irq.c ? Or we just remove the dependency with no code relocation?

Reply:
@amr-sc possibly? Please, make a sensible choice about this topics and make a proposal, we can discuss on the code. If I had all the answers I would write the code myself.

Reply:
@carlocaione , done